controllers/oauth2client_controller.go

```diff
@@ -18,46 +18,58 @@ package controllers
 import (
 	"context"
 	"fmt"
+
 	"github.com/go-logr/logr"
-	hydrav1alpha1 "github.com/ory/hydra-maester/api/v1alpha1"
-	"github.com/ory/hydra-maester/hydra"
 	"github.com/pkg/errors"
 	apiv1 "k8s.io/api/core/v1"
 	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hydrav1alpha1 "github.com/ory/hydra-maester/api/v1alpha1"
+	"github.com/ory/hydra-maester/hydra"
 )
 
 const (
 	ClientIDKey     = "client_id"
 	ClientSecretKey = "client_secret"
 	FinalizerName   = "finalizer.ory.hydra.sh"
+
+	DefaultNamespace = "default"
 )
 
-type clientMapKey struct {
+type clientKey struct {
 	url            string
 	port           int
 	endpoint       string
 	forwardedProto string
 }
 
-type HydraClientInterface interface {
-	GetOAuth2Client(id string) (*hydra.OAuth2ClientJSON, bool, error)
-	ListOAuth2Client() ([]*hydra.OAuth2ClientJSON, error)
-	PostOAuth2Client(o *hydra.OAuth2ClientJSON) (*hydra.OAuth2ClientJSON, error)
-	PutOAuth2Client(o *hydra.OAuth2ClientJSON) (*hydra.OAuth2ClientJSON, error)
-	DeleteOAuth2Client(id string) error
-}
-
-// OAuth2ClientReconciler reconciles a OAuth2Client object
+// OAuth2ClientReconciler reconciles a OAuth2Client object.
 type OAuth2ClientReconciler struct {
-	HydraClient  HydraClientInterface
-	Log          logr.Logger
-	otherClients map[clientMapKey]HydraClientInterface
 	client.Client
+	HydraClient         hydra.Client
+	Log                 logr.Logger
 	ControllerNamespace string
+
+	oauth2Clients map[clientKey]hydra.Client
```

Comment:
IMO we don't need to have a) default hydra client (`HydraClient`) and b) a list of clients.
We can have just this list and take the default one from there. 

In general, I also don't think we need this list (I believe it was a form of cache originally) but your PR is trying to fix broken logic so you don't have to remove this list it in your PR ;)

Reply:
I figured it was meant to be used as some kind of cache for instantiated oauth2 clients, although I'm not too sure if it's really useful because that can only help for different custom resources using the same parameters. Might it be better to cache from the custom resource name? Proper invalidation will be needed during finalizer phase though.

Reply:
Caching based on CRname would increase cache size when you have for ex multiple CRs in same or different NS but using same hydra service. It's not critical for now but I think in the future we will need to refactor it.